tiledb/api/src/query/write/input/arrow.rs

```diff
@@ -0,0 +1,753 @@
+use std::num::NonZeroU32;
+use std::rc::Rc;
+use std::sync::Arc;
+
+use anyhow::anyhow;
+use arrow::array::{
+    Array as ArrowArray, AsArray, FixedSizeBinaryArray, FixedSizeListArray,
+    GenericListArray, LargeBinaryArray, LargeStringArray, PrimitiveArray,
+    RecordBatch,
+};
+use arrow::buffer::OffsetBuffer;
+use arrow::datatypes::{ArrowPrimitiveType, Field};
+
+use crate::array::{CellValNum, Schema};
+use crate::datatype::PhysicalType;
+use crate::error::{DatatypeErrorKind, Error};
+use crate::query::buffer::{
+    Buffer, CellStructure, QueryBuffers, TypedQueryBuffers,
+};
+use crate::query::write::input::{
+    DataProvider, RecordProvider, TypedDataProvider,
+};
+use crate::Result as TileDBResult;
+
+fn cell_structure_var(
+    offsets: &OffsetBuffer<i64>,
+    cell_val_num: CellValNum,
+) -> TileDBResult<CellStructure> {
+    match cell_val_num {
+        CellValNum::Fixed(nz) => {
+            let expect_len = nz.get() as i64;
+            for window in offsets.windows(2) {
+                if window[1] - window[0] != expect_len {
+                    return Err(Error::Datatype(
+                        DatatypeErrorKind::UnexpectedCellStructure {
+                            context: Some(
+                                "Arrow array as query input".to_owned(),
+                            ),
+                            expected: cell_val_num,
+                            found: CellValNum::Var,
+                        },
+                    ));
+                }
+            }
+            Ok(CellStructure::Fixed(nz))
+        }
+        CellValNum::Var => Ok(CellStructure::Var(Buffer::Borrowed(
+            &offsets.inner().inner().typed_data::<u64>()[0..offsets.len() - 1],
+        ))),
+    }
+}
+
+fn cell_structure_fixed(
+    fixed_len: i32,
+    ncells: usize,
+    cell_val_num: CellValNum,
+) -> TileDBResult<CellStructure<'static>> {
+    match cell_val_num {
+        CellValNum::Fixed(nz) if fixed_len as u32 != nz.get() => Err(
+            Error::Datatype(DatatypeErrorKind::UnexpectedCellStructure {
+                context: Some("Arrow array as query input".to_owned()),
+                expected: cell_val_num,
+                found: match NonZeroU32::new(fixed_len as u32) {
+                    Some(nz) => CellValNum::Fixed(nz),
+                    None => CellValNum::Var,
+                },
+            }),
+        ),
+        CellValNum::Fixed(nz) => Ok(CellStructure::Fixed(nz)),
+        CellValNum::Var => {
+            let offsets = Buffer::Owned(
+                std::iter::repeat(fixed_len as usize)
+                    .take(ncells)
+                    .enumerate()
+                    .map(|(i, len)| (i * len) as u64)
+                    .collect::<Vec<u64>>()
+                    .into_boxed_slice(),
+            );
```

Comment:
Not asking for a change here, but did you try looking at the assembly or one of the IR outputs to get an idea on the efficiency for this? I'm curious if going the other route wouldn't be faster. Something like:

```rust
let data = Vec::<u64>::new();
data.iter_mut().enumerate().for_each(|idx, val| *val = idx * len);
```

I'm mostly just wondering if one or the other would get pushed down to a single allocation with a single auto vectorized loop for initialization or not.

Reply:
Haven't done that yet.  Seems like a good exercise.

Reply:
I don't remember what I was doing exactly, but it was something like making sure that something I did was either a memcpy or memset instead of iterating over an entire vector. If memory serves there's a blog post or SO post that says what compiler output level to look at to see if its inserting a syscall or not. I'm not entirely sure how to check for auto vectorization.

Reply:
The version you offer is fewer instructions and looks simpler so I'll pivot to that

Reply:
Cool beans!